### PR TITLE
feat: Configure Loadgen Options

### PR DESCRIPTION
**Pull request by AbhiPrasad**

### CODE REVIEW

**Comment by AbhiPrasad** [comment]
Will update result template after https://github.com/getsentry/sentry-sdk-benchmark/pull/12 gets merged in.

File: tool/loadgen/main.go [hunk 1]
```diff
@@ -37,20 +45,33 @@ func main() {
...
 
+	result := run(options)
 	save(result, path)
 }
 
+func run(options LoadGenOptions) TestResult {
+	waitUntilReady(options.WaitGen)
+	warmUp(options.WarmupGen)
+	result := test(options.ContainerName, options.TestGen)
+	result.Options = options
```
Comment (AbhiPrasad) [inline tool/loadgen/main.go:64]: Not fully comfortable with this pattern, but didn't think too hard about it

File: tool/loadgen/main.go [hunk 2]
```diff
@@ -18,6 +18,14 @@ const target = "http://app:8080"
 
 var hasRelay = os.Getenv("HAS_RELAY") == "true"
 
+type LoadGenOptions struct {
+	Path          string    `json:"path"`
+	ContainerName string    `json:"container_name"`
+	WaitGen       Generator `json:"wait_generator"`
```
Comment (rhcarvalho) [inline tool/loadgen/main.go:24]: Perhaps the "wait" phase can be ignored from the parametrization? It does affect what Relay sees, and eventually what goes into the result, but if we wanted we could somehow exclude it from results.

The thing is this steps is only there to ensure the server is up and ready to accept requests.

File: tool/loadgen/main.go [hunk 3]
```diff
@@ -126,16 +148,32 @@ func test(containerName string) TestResult {
 	return tr
 }
 
+type Generator struct {
+	Url      string        `json:"url"`
+	Duration time.Duration `json:"duration"`
+	Rate     vegeta.Rate   `json:"rate"`
+}
+
+func NewGenerator(url string, duration time.Duration) Generator {
```
Comment (rhcarvalho) [inline tool/loadgen/main.go:157]: It is not clear that we'll always use the same rate for all phases. Have you considered explicitly passing all options when creating a `Generator` value (i.e. no constructor function)?

File: tool/loadgen/main.go [hunk 4]
```diff
@@ -18,6 +18,14 @@ const target = "http://app:8080"
 
 var hasRelay = os.Getenv("HAS_RELAY") == "true"
 
+type LoadGenOptions struct {
+	Path          string    `json:"path"`
+	ContainerName string    `json:"container_name"`
```
Comment (rhcarvalho) [inline tool/loadgen/main.go:23]: Did you find this to be useful? It's going to be a string like `django-baseline-jio6rry`. With a quick check I think this string is not to be seen in the result, but appears in the (debug) compose file.

**Review by rhcarvalho — COMMENTED** [review]
Good stuff 👍 

A couple points to discuss.

File: tool/loadgen/main.go (hunk 4 above)
Comment (AbhiPrasad) [reply tool/loadgen/main.go]: I pulled it in because it was set from an env variable, but you're right that it doesn't provide much value. Will remove.

**Comment by AbhiPrasad** [comment]
Should I apply this patch?

```diff
diff --git a/tool/loadgen/main.go b/tool/loadgen/main.go
index 36456a3..9dd04de 100644
--- a/tool/loadgen/main.go
+++ b/tool/loadgen/main.go
@@ -49,6 +49,7 @@ func main() {
 		Per:  time.Second,
 	}
 	url := target + "/update?query=100"
+	method := "GET"
 
 	options := LoadGenOptions{
 		Path: path,
@@ -56,11 +57,13 @@ func main() {
 			Url:      url,
 			Duration: 15 * time.Second,
 			Rate:     rate,
+			Method:   method,
 		},
 		TestGen: Generator{
 			Url:      url,
 			Duration: 20 * time.Second,
 			Rate:     rate,
+			Method:   method,
 		},
 	}
 
@@ -68,6 +71,7 @@ func main() {
 		Url:      target + "/update?query=1",
 		Duration: 5 * time.Second,
 		Rate:     rate,
+		Method:   method,
 	})
 	result := run(containerName, options)
 	save(result, path)
@@ -166,13 +170,14 @@ type Generator struct {
 	Url      string        `json:"url"`
 	Duration time.Duration `json:"duration"`
 	Rate     vegeta.Rate   `json:"rate"`
+	Method   string        `json:"method"`
 }
 
 // fetch requests the given URL several times for the given duration and with
 // the given concurrency level.
 func (g *Generator) fetch() *vegeta.Metrics {
 	target := vegeta.NewStaticTargeter(vegeta.Target{
-		Method: "GET",
+		Method: g.Method,
 		URL:    g.Url,
 	})
 	attacker := vegeta.NewAttacker()
```

**Comment by rhcarvalho** [comment]
> Should I apply this patch?

Talked; let's limit the scope to only supporting GET. Other verbs would potentially require passing data or other arrangements, requiring further changes.

**Comment by AbhiPrasad** [comment]
Parametrizing the run doesn’t seem useful at the moment. So we’re closing this as won’t do.

**Pull request closed without merging**